src/net/echo.zig

```diff
@@ -51,204 +44,306 @@ pub fn getShredAndIPFromEchoServer(
+        try request.reader(),
+        std.http.Server.Request.ReadError,
+    );
 
-pub fn requestIpEcho(
-    allocator: std.mem.Allocator,
-    addr: std.net.Address,
-    message: IpEchoServerMessage,
-) !IpEchoServerResponse {
-    // connect + send
-    const conn = try std.net.tcpConnectToAddress(addr);
-    defer conn.close();
-    try conn.writeAll(&(.{0} ** HEADER_LENGTH));
-    try bincode.write(conn.writer(), message, .{});
-    try conn.writeAll("\n");
-
-    // get response
-    var buff: [32]u8 = undefined;
-    const len = try conn.readAll(&buff);
-    var bufferStream = std.io.fixedBufferStream(buff[HEADER_LENGTH..len]);
-    return try bincode.read(allocator, IpEchoServerResponse, bufferStream.reader(), .{});
-}
-
-// TODO: FIX ME
-// test "net.echo: Server works" {
-//     const port: u16 = 34333;
-
-//     var exit = Atomic(bool).init(false);
-
-//     var server = Server.init(testing.allocator, port, &exit);
-//     defer server.deinit();
-
-//     var server_thread_handle = try server.listenAndServe();
-//     if (builtin.os.tag == .linux) try server_thread_handle.setName("server_thread");
+    if (request.head.method != .POST or
+        !std.mem.eql(u8, request.head.target, "/"))
+    {
+        try httpRespondError(&request, .not_found, "");
+        return;
+    }
 
-//     var client = std.http.Client{ .allocator = testing.allocator };
-//     defer client.deinit();
+    hct.logger.debug().log("handling echo request");
 
-//     var server_header_buff = [_]u8{0} ** 1024;
+    var request_buf: [256]u8 = undefined;
+    const request_byte_count = try reader.readAll(&request_buf);
+    const request_bytes = request_buf[0..request_byte_count];
 
-//     // create request
-//     var req = try client.open(.POST, try std.Uri.parse("http://localhost:34333/"), .{
-//         .server_header_buffer = &server_header_buff,
-//         .headers = .{
-//             .content_type = .{ .override = "text/plain" },
-//             // .accept_encoding = .{ .override = "*/*" },
-//         },
-//         .extra_headers = &.{
-//             // .{
-//             //     .name = "connection",
-//             //     .value = "close",
-//             // },
-//         },
-//     });
-//     defer req.deinit();
+    const ip_echo_request_result = std.json.parseFromSlice(
+        IpEchoRequest,
+        hct.allocator,
+        request_bytes,
+        .{ .allocate = .alloc_if_needed },
+    ) catch |err| return switch (err) {
+        error.OutOfMemory => try httpRespondError(&request, .insufficient_storage, ""),
+        else => try httpRespondError(&request, .bad_request, ""),
+    };
+    const ip_echo_request = ip_echo_request_result.value;
+    _ = ip_echo_request;
+    ip_echo_request_result.deinit();
+
+    const socket_addr = SocketAddr.fromIpV4Address(request.server.connection.address);
+    const ip_echo_response: IpEchoResponse = .{
+        .address = .{ .ipv4 = socket_addr.V4.ip },
+        .shred_version = .{ .value = 0 },
```

Comment:
we should have a TODO or ticket that the correct shred version should be propagated here

Reply:
d5b2dcb854dacc5ce211d6251c05953c8ed79705